src/node/channels.h

```diff
@@ -749,31 +751,35 @@ namespace ccf
         return false;
       }
 
-      RecvNonce nonce(
-        send_nonce.fetch_add(1), threading::get_current_thread_id());
-
-      GcmHdr gcm_hdr;
-      gcm_hdr.set_iv_seq(nonce.get_val());
-
       assert(send_key);
 
       // During key rollover, we keep recv_key to decrypt messages from the peer
       // until it has rolled over too (recognized when we received a message
       // with the nonce/seqno reset to 1). But, we can immediately start to send
       // messages with the new send_key.
 
+      GcmHdr gcm_hdr;
+      RecvNonce nonce(
+        send_nonce.fetch_add(1), threading::get_current_thread_id());
+      gcm_hdr.set_iv_seq(nonce.get_val());
+
       std::vector<uint8_t> cipher(plain.n);
       send_key->encrypt(
         gcm_hdr.get_iv(), plain, aad, cipher.data(), gcm_hdr.tag);
 
-      to_host->write(
-        node_outbound,
-        peer_id.value(),
-        type,
-        self.value(),
-        serializer::ByteRange{aad.p, aad.n},
-        gcm_hdr,
-        cipher);
+      const auto gcm_hdr_serialised = gcm_hdr.serialise();
+
+      // Payload is concatenation of 3 things:
+      // 1) aad
+      // 2) gcm header
+      // 3) ciphertext
+      const serializer::ByteRange payload[] = {
+        {aad.p, aad.n},
+        {gcm_hdr_serialised.data(), gcm_hdr_serialised.size()},
+        {cipher.data(), cipher.size()}};
+
+      RINGBUFFER_WRITE_MESSAGE(
+        node_outbound, to_host, peer_id.value(), type, self.value(), payload);
```

Comment:
To make this flow explicit:

A `node_outbound` message is a ringbuffer message that contains 4 fields: `Peer`, `Type`, `Self`, `Payload`.
When the host reads that, it reads the first 3 (because if `Type` has the correct value, it may append an additional payload from the ledger).
Using `Peer` as the key, the host then sends the other 3 to the intended node. This could be serialised or retransmitted in some other carrier format! In practice we reuse the existing serialised form and blit it over TCP directly, but this is a totally distinct transfer.
On the receiving host, we see that incoming message and read the fields in it: `Type`, `From`, `Payload`. Those are written as `node_inbound` over the ringbuffer.

We previously benefitted from several places where luckily the serialisation methods matched up, and I think they're now more explicit.